tests/unit/s2n_send_test.c

```diff
@@ -124,18 +246,360 @@ int main(int argc, char **argv)
+         * 1. actual_protocol_version < S2N_TLS11 
+         * 2. connection is in client mode
+         * 3. the chosen cipher uses CBC */
+        conn->actual_protocol_version = S2N_TLS10;
+        /* This can be any cipher suite that uses CBC. s2n_rsa_with_3des_ede_cbc_sha is chosen because it's easier to stub. */
+        struct s2n_cipher_suite *cipher_suite = &s2n_rsa_with_3des_ede_cbc_sha;
+        struct s2n_cipher_suite *composite_cipher_suite = &s2n_rsa_with_aes_128_cbc_sha;
+        conn->client->cipher_suite = cipher_suite;
+
+        /* We need to check out if a composite cipher is available to filter out BoringSSL and AWSLC, or s2n_send will fail trying to use CBC. */
+        if (cipher_suite->available && composite_cipher_suite->all_record_algs[0]->cipher->is_available()) {
+            s2n_custom_send_fn_called = false;
+            EXPECT_SUCCESS(s2n_send(conn, test_data, sizeof(test_data), &blocked));
+            EXPECT_TRUE(s2n_custom_send_fn_called);
+            EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+        }
+    }
+
+    /* s2n_send partial writes are flushed in proceeding sends */
+    {
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+            s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        uint32_t actual_bytes_written = 0;
+        EXPECT_SUCCESS(s2n_connection_set_ctx(conn, (void*)&actual_bytes_written));
+
+        s2n_blocked_status blocked = 0;
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_partial_socket_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+        s2n_custom_send_fn_called = false;
+
+        /* Assume we have a partial write of sizeof(test_data) - 1 that was successful.
+         * The subsequent socket write failed with EAGAIN, and control has returned to the application.*/
+        EXPECT_FAILURE_WITH_ERRNO(s2n_send(conn, test_data, sizeof(test_data), &blocked), S2N_ERR_IO_BLOCKED);
+        EXPECT_EQUAL(conn->current_user_data_consumed, sizeof(test_data));
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_WRITE);
+
+        /* This is to be kept in sync with the partial write performed by s2n_partial_socket_send_fn.
+         * As of this comment it will perform the entire right except the last 3 bytes. */
+        uint32_t expected_remaining_data = sizeof(test_data) - (sizeof(test_data) - 3);
+        EXPECT_EQUAL(s2n_stuffer_data_available(&conn->out), expected_remaining_data);
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_WRITE);
+
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_send_always_passes_fn));
+        EXPECT_SUCCESS(s2n_send(conn, test_data, sizeof(test_data), &blocked));
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+        EXPECT_EQUAL(conn->current_user_data_consumed, 0);
+        EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
```

Comment:
How does this prove that we flushed the previous partial write?

Reply:
I assumed `EXPECT_EQUAL(conn->current_user_data_consumed, 0);` would be enough, but you are right! I will add another check that there is no data available in the out stuffer.